domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt

```diff
@@ -161,101 +208,340 @@ class TopicListController @Inject constructor(
-
+        val completedChapterProgressList = getCompletedChapterProgressList(storyProgress)
         val lastCompletedChapterProgress: ChapterProgress? =
           completedChapterProgressList.firstOrNull()
 
-        val startedChapterProgressList =
-          storyProgress.chapterProgressMap.values
-            .filter { chapterProgress ->
-              chapterProgress.chapterPlayState ==
-                ChapterPlayState.STARTED_NOT_COMPLETED
-            }
-            .sortedByDescending { chapterProgress -> chapterProgress.lastPlayedTimestamp }
+        val startedChapterProgressList = getStartedChapterProgressList(storyProgress)
 
         val recentlyPlayerChapterProgress: ChapterProgress? =
           startedChapterProgressList.firstOrNull()
-        if (recentlyPlayerChapterProgress != null) {
-          val recentlyPlayerChapterSummary: ChapterSummary? =
-            story.chapterList.find { chapterSummary ->
-              recentlyPlayerChapterProgress.explorationId == chapterSummary.explorationId
-            }
-          if (recentlyPlayerChapterSummary != null) {
-            val numberOfDaysPassed =
-              (Date().time - recentlyPlayerChapterProgress.lastPlayedTimestamp) / ONE_DAY_IN_MS
-            val promotedStory = createPromotedStory(
-              storyId,
-              topic,
-              completedChapterProgressList.size,
-              story.chapterCount,
-              recentlyPlayerChapterSummary.name,
-              recentlyPlayerChapterSummary.explorationId
-            )
-            if (numberOfDaysPassed < ONE_WEEK_IN_DAYS) {
-              ongoingStoryListBuilder.addRecentStory(promotedStory)
-            } else {
-              ongoingStoryListBuilder.addOlderStory(promotedStory)
+
+        when {
+          recentlyPlayerChapterProgress != null -> {
+            val recentlyPlayerChapterSummary: ChapterSummary? =
+              story.chapterList.find { chapterSummary ->
+                recentlyPlayerChapterProgress.explorationId == chapterSummary.explorationId
+              }
+            if (recentlyPlayerChapterSummary != null) {
+              val numberOfDaysPassed = (
+                oppiaClock.getCurrentCalendar().timeInMillis -
+                  recentlyPlayerChapterProgress.lastPlayedTimestamp
+                ) / ONE_DAY_IN_MS
+              val promotedStory = createPromotedStory(
+                storyId,
+                topic,
+                completedChapterProgressList.size,
+                story.chapterCount,
+                recentlyPlayerChapterSummary.name,
+                recentlyPlayerChapterSummary.explorationId
+              )
+              if (numberOfDaysPassed < ONE_WEEK_IN_DAYS) {
+                recommendedStoryBuilder.addRecentlyPlayedStory(promotedStory)
+              } else {
+                recommendedStoryBuilder.addOlderPlayedStory(promotedStory)
+              }
```

Comment:
This & below also seems a bit redundant with code above & below. Can we further consolidate some of this logic to avoid repeating the same logic, if statements, etc.?

Reply:
Done.